syntax = "proto3";

package game;

service User {
    rpc UsernameAvailable(UserDetails) returns (UserDetailsResponse) {}
    rpc UsernameValid(UserDetails) returns (UserDetailsResponse) {}
    rpc EmailValid(UserDetails) returns (UserDetailsResponse) {}
    rpc CreateUserDetails(UserDetails) returns (UserDetails) {}  
    rpc GetUserDetails(UserDetails) returns (UserDetails) {}
    rpc UpdateUserDetails(UserDetails) returns (UserDetails) {}
    rpc GetStatistics(StatisticsRequest) returns (stream GameStatistic) {}
}

message UserDetailsResponse {
    bool result = 1;
}

message UserDetails {
    uint32 id = 1;
    string username = 2;
    string email = 3;
}

message StatisticsRequest {
    uint32 user_id = 1;
    uint32 game_id = 2;
}

message GameStatistic {
    uint32 user_id = 1;
    uint32 game_id = 2;
    string name = 3;
    string value = 4;
    enum DataType {
        STR = 0;
        INT = 1;
        FLOAT = 2;
        BOOL = 3;
    } 
    DataType type = 5;
}

message Achievement {
//user_id
//game_id
//desc
//name
//resource 
}

message Match {
//game_id
//[user_d] (participants)
}
